core/trino-spi/src/main/java/io/trino/spi/type/TypeOperators.java

```diff
@@ -69,7 +69,15 @@
     public TypeOperators()
     {
         ConcurrentHashMap<Object, Object> cache = new ConcurrentHashMap<>();
-        this.cache = (operatorConvention, supplier) -> cache.computeIfAbsent(operatorConvention, key -> supplier.get());
+        // optimize the read lock caused by computeIfAbsent
+        // this.cache = (operatorConvention, supplier) -> cache.computeIfAbsent(operatorConvention, key -> supplier.get());
+        this.cache = (operatorConvention, supplier) -> {
+            Object value = cache.get(operatorConvention);
+            if (value == null) {
+                value = cache.computeIfAbsent(operatorConvention, key -> supplier.get());
```

Comment:
according to doc:
```
Some attempted update
     * operations on this map by other threads may be blocked while
     * computation is in progress, so the computation should be short
     * and simple.
```
looking at `java.util.concurrent.ConcurrentHashMap#computeIfAbsent` it shouldn't lock if element is already present, so once cache is filled, there should be no locking.

Why do you think this change matters in non-benchmark executions

Reply:
During large-scale data testing, we found that this code has performance problems. When modifying, we also refer to the transformation idea of computeifabsent method of mybatis. The following is the modification of mybatis:

https://github.com/mybatis/mybatis-3/pull/2223

Reply:
> During large-scale data testing, we found that this code has performance problems.

How did you do the testing?
`io.trino.spi.type.TypeOperators#cache` should not be changed after a while since all operators would be cached. Hence, there should be no locking